Custom Component compile error when update from 1.12.0 to 1.13.5

<!-- Thanks for reporting a bug for this project. READ THIS FIRST:

- Provide as many details as possible. Simply saying "X gives bug" or "X gives error" is not enough!
- Paste logs, configuration sample and code into the backticks (```).
- Read through the template carefully and fill out all missing details.
- Please also search for similar issues in this issue tracker first and read through the ESPHome FAQ.

DO NOT DELETE ANY TEXT from this template! Otherwise the issue may be closed without a comment.
-->

**Operating environment/Installation (Hass.io/Docker/pip/etc.):**
<!--
Please provide details about your environment below this line. -->

**ESP (ESP32/ESP8266, Board/Sonoff):**
<!--
Please provide details about which ESP you're using below.
-->ESP01-s
**Affected component:**
<!--
Please add the link to the documentation at https://esphome.io/index.html of the component in question.
-->
UARTDevice, mqtt::CustomMQTTDevice

**Description of problem:**
in tonecommand.h:
#include "esphome.h"

// part to asr 
class ToneCommand : public Component, public UARTDevice, public mqtt::CustomMQTTDevice {
  public:
    ToneCommand(UARTComponent *parent) : UARTDevice(parent) {}
    
    sensor::Sensor *index = new sensor::Sensor();
    text_sensor::TextSensor *service = new text_sensor::TextSensor();
    text_sensor::TextSensor *entity_id = new text_sensor::TextSensor();
    
    void setup() override {
      write_array(txdata[0],3); // asr init
      index->publish_state(0);

      subscribe("tone/command", &ToneCommand::on_message);
      subscribe_json("tone/json/command", &ToneCommand::on_json_message);
    }
    void loop() override {
      if(available()>0){
        delay(50);
        numdata = readBytes(buffer,3);
    ......


**Problem-relevant YAML-configuration entries:**
```yaml
esphome:
  includes:
    - tonecommand.h
  name: tone_command
  platform: ESP8266
  board: esp01_1m

logger:
  baud_rate: 0

uart:
  - id: utonecommand
    tx_pin: GPIO1
    rx_pin: GPIO3
    baud_rate: 115200
  - id: utonetts
    tx_pin: GPIO2
    baud_rate: 9600

wifi:
  networks:
  - ssid: !secret wifi_ssid
    password: !secret wifi_password
  - ssid: !secret wifi_ssid1
    password: !secret wifi_password

ota:
  password: !secret ota_password

api:
#  password: !secret api_password
  services:
    - service: tonecommand
      variables:
        command: string
      then:
        - mqtt.publish:
            topic: 'tone/command'
            payload: !lambda 'return command;'

mqtt:
  broker: !secret mqtt_broker
  username: !secret mqtt_username
  password: !secret mqtt_password
  discovery: True
  discovery_retain: true

switch:
  - platform: uart
    name: "tone continue"
    id: stonecontinue
    uart_id: utonecommand
    data: [0xAB,0xAB,0x00]

  - platform: uart
    name: "tone tts"
    id: stonetts
    uart_id: utonetts
    data: '<I>18'

binary_sensor:
  - platform: gpio
    pin: GPIO0
    id: btntonecontrol
    name: "btnToneControl"
    internal: true
    filters:
      - invert:
    on_multi_click: 
    - timing: #Double Clicked
        - ON for at most 1s
        - OFF for at most 1s
        - ON for at most 1s
        - OFF for at least 0.2s
      then:
        - switch.turn_on:
            id: stonestop
        - switch.turn_on:
            id: stonesleep
    - timing: #Single Long Clicked
        - ON for 3s to 5s
        - OFF for at least 0.5s
      then:
        - switch.turn_on:
            id: stonestop
        - delay: 00:00:01
        - switch.turn_on:
            id: stonecontinue
    - timing: #Single Short Clicked
        - ON for at most 2s
        - OFF for at least 0.5s
      then:
        - switch.turn_on:
            id: stonestop
        - delay: 00:00:01
        - switch.turn_on:
            id: stoneonce

sensor:
  - platform: wifi_signal
    name: "WiFi Signal_tone"
    update_interval: 60s
  - platform: custom
    lambda: |-
      auto tonecommand = new ToneCommand(id(utonecommand));
      App.register_component(tonecommand);
      return {tonecommand->index};
    sensors:
      - name: "command index"
        id: "commandindexid"

text_sensor:
  - platform: custom
    lambda: |-
      auto my_tone_command = new ToneCommand(id(utonecommand));
      App.register_component(my_tone_command);
      return {my_tone_command->service,my_tone_command->entity_id};
    text_sensors:
      - name: "tone service"
        id: "toneservice"
      - name: "tone entity_id"
        id: "tonentityid"
 ......
```

**Traceback (if applicable):**
<!--
Please copy the traceback here if compilation is failing. If possible, also connect to the ESP over USB and copy its logs into the backticks.
-->
```
Compiling .pioenvs/tone_command/libd05/ESP8266WiFi/WiFiClient.cpp.o
In file included from src/main.cpp:58:0:
src/tonecommand.h:4:56: error: expected class-name before ',' token
class ToneCommand : public Component, public UARTDevice, public MQTTClientComponent {
^
src/tonecommand.h:6:31: error: expected ')' before '*' token
ToneCommand(UARTComponent *parent) : UARTDevice(parent) {}
^
src/tonecommand.h:4:7: warning: direct base 'esphome::Component' inaccessible in 'ToneCommand' due to ambiguity [enabled by default]
class ToneCommand : public Component, public UARTDevice, public MQTTClientComponent {
^
src/tonecommand.h: In member function 'virtual void ToneCommand::setup()':
src/tonecommand.h:13:30: error: 'write_array' was not declared in this scope
write_array(txdata[0],3); // asr init
^
src/tonecommand.h: In member function 'virtual void ToneCommand::loop()':
src/tonecommand.h:20:20: error: 'available' was not declared in this scope
if(available()>0){
 ......
```

**Additional information and things you've tried:**

I tried examples in docs(custom UARTDevice), found same errors too, but in 1.12.1, 
it's none error and work fine

<!-- LEAVE THIS LINE AS-IS AND DON'T DELETE IT, OTHERWISE THE ISSUE WILL BE CLOSED AUTOMATICALLY. -->

Not sure if it's related, and I hope it can help, but I have the same issue with the custom component for [MAX7219 in dot grid config](https://github.com/ASMfreaK/esphome_max7219)

When compiling on 1.13.5 : 
````
In file included from src/main.cpp:10:0:
src/max7219grid.hpp:1:29: fatal error: esphome/defines.h: No such file or directory
compilation terminated.
*** [/data/reveil/.pioenvs/reveil/src/main.cpp.o] Error 1
````

It does compile on 1.12.2

--- a/esphome/components/mqtt/custom_mqtt_device.cpp
+++ b/esphome/components/mqtt/custom_mqtt_device.cpp
@@ -1,0 +1,30 @@
+#include "custom_mqtt_device.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace mqtt {
+
+static const char *TAG = "mqtt.custom";
+
+bool CustomMQTTDevice::publish(const std::string &topic, const std::string &payload, uint8_t qos, bool retain) {
+  return global_mqtt_client->publish(topic, payload, qos, retain);
+}
+bool CustomMQTTDevice::publish(const std::string &topic, float value, int8_t number_decimals) {
+  auto str = value_accuracy_to_string(value, number_decimals);
+  return this->publish(topic, str);
+}
+bool CustomMQTTDevice::publish(const std::string &topic, int value) {
+  char buffer[24];
+  sprintf(buffer, "%d", value);
+  return this->publish(topic, buffer);
+}
+bool CustomMQTTDevice::publish_json(const std::string &topic, const json_build_t &f, uint8_t qos, bool retain) {
+  return global_mqtt_client->publish_json(topic, f, qos, retain);
+}
+bool CustomMQTTDevice::publish_json(const std::string &topic, const json_build_t &f) {
+  return this->publish_json(topic, f, 0, false);
+}
+bool CustomMQTTDevice::is_connected() { return global_mqtt_client != nullptr && global_mqtt_client->is_connected(); }
+
+}  // namespace mqtt
+}  // namespace esphome

--- a/esphome/components/mqtt/custom_mqtt_device.cpp
+++ b/esphome/components/mqtt/custom_mqtt_device.cpp
@@ -18,10 +18,10 @@ bool CustomMQTTDevice::publish(const std::string &topic, int value) {
   sprintf(buffer, "%d", value);
   return this->publish(topic, buffer);
 }
-bool CustomMQTTDevice::publish_json(const std::string &topic, const json_build_t &f, uint8_t qos, bool retain) {
+bool CustomMQTTDevice::publish_json(const std::string &topic, const json::json_build_t &f, uint8_t qos, bool retain) {
   return global_mqtt_client->publish_json(topic, f, qos, retain);
 }
-bool CustomMQTTDevice::publish_json(const std::string &topic, const json_build_t &f) {
+bool CustomMQTTDevice::publish_json(const std::string &topic, const json::json_build_t &f) {
   return this->publish_json(topic, f, 0, false);
 }
 bool CustomMQTTDevice::is_connected() { return global_mqtt_client != nullptr && global_mqtt_client->is_connected(); }

--- a/esphome/components/mqtt/custom_mqtt_device.h
+++ b/esphome/components/mqtt/custom_mqtt_device.h
@@ -1,0 +1,217 @@
+#pragma once
+
+#include "esphome/core/component.h"
+#include "mqtt_client.h"
+
+namespace esphome {
+namespace mqtt {
+
+/** This class is a helper class for custom components that communicate using
+ * MQTT. It has 5 helper functions that you can use (square brackets indicate optional):
+ *
+ *  - `subscribe(topic, function_pointer, [qos])`
+ *  - `subscribe_json(topic, function_pointer, [qos])`
+ *  - `publish(topic, payload, [qos], [retain])`
+ *  - `publish_json(topic, payload_builder, [qos], [retain])`
+ *  - `is_connected()`
+ */
+class CustomMQTTDevice {
+ public:
+  /** Subscribe to an MQTT topic with the given Quality of Service.
+   *
+   * Example:
+   *
+   * ```cpp
+   * class MyCustomMQTTDevice : public Component, public mqtt:CustomMQTTDevice {
+   *  public:
+   *   void setup() override {
+   *     subscribe("the/topic", &MyCustomMQTTDevice::on_message);
+   *     pinMode(5, OUTPUT);
+   *   }
+   *
+   *   // topic and payload parameters can be removed if not needed
+   *   // e.g: void on_message() {
+   *
+   *   void on_message(const std::string &topic, const std::string &payload) {
+   *     // do something with topic and payload
+   *     if (payload == "ON") {
+   *       digitalWrite(5, HIGH);
+   *     } else {
+   *       digitalWrite(5, LOW);
+   *     }
+   *   }
+   * };
+   * ```
+   *
+   * @tparam T A C++ template argument for determining the type of the callback.
+   * @param topic The topic to subscribe to. Re-subscription on re-connects is automatically handled.
+   * @param callback The callback (must be a class member) to subscribe with.
+   * @param qos The Quality of Service to subscribe with. Defaults to 0.
+   */
+  template<typename T>
+  void subscribe(const std::string &topic, void (T::*callback)(const std::string &, const std::string &),
+                 uint8_t qos = 0);
+
+  template<typename T>
+  void subscribe(const std::string &topic, void (T::*callback)(const std::string &), uint8_t qos = 0);
+
+  template<typename T> void subscribe(const std::string &topic, void (T::*callback)(), uint8_t qos = 0);
+
+  /** Subscribe to an MQTT topic and call the callback if the payload can be decoded
+   * as JSON with the given Quality of Service.
+   *
+   * Example:
+   *
+   * ```cpp
+   * class MyCustomMQTTDevice : public Component, public mqtt:CustomMQTTDevice {
+   *  public:
+   *   void setup() override {
+   *     subscribe_json("the/topic", &MyCustomMQTTDevice::on_json_message);
+   *     pinMode(5, OUTPUT);
+   *   }
+   *
+   *   // topic parameter can be remove if not needed:
+   *   // e.g.: void on_json_message(JsonObject &payload) {
+   *
+   *   void on_json_message(const std::string &topic, JsonObject &payload) {
+   *     // do something with topic and payload
+   *     if (payload["number"] == 1) {
+   *       digitalWrite(5, HIGH);
+   *     } else {
+   *       digitalWrite(5, LOW);
+   *     }
+   *   }
+   * };
+   * ```
+   *
+   * @tparam T A C++ template argument for determining the type of the callback.
+   * @param topic The topic to subscribe to. Re-subscription on re-connects is automatically handled.
+   * @param callback The callback (must be a class member) to subscribe with.
+   * @param qos The Quality of Service to subscribe with. Defaults to 0.
+   */
+  template<typename T>
+  void subscribe_json(const std::string &topic, void (T::*callback)(const std::string &, JsonObject &),
+                      uint8_t qos = 0);
+
+  template<typename T>
+  void subscribe_json(const std::string &topic, void (T::*callback)(JsonObject &), uint8_t qos = 0);
+
+  /** Publish an MQTT message with the given payload and QoS and retain settings.
+   *
+   * Example:
+   *
+   * ```cpp
+   * void in_some_method() {
+   *   publish("the/topic", "The Payload", 0, true);
+   * }
+   * ```
+   *
+   * @param topic The topic to publish to.
+   * @param payload The payload to publish.
+   * @param qos The Quality of Service to publish with. Defaults to 0
+   * @param retain Whether to retain the message. Defaults to false.
+   */
+  bool publish(const std::string &topic, const std::string &payload, uint8_t qos = 0, bool retain = false);
+
+  /** Publish an MQTT message with the given floating point number and number of decimals.
+   *
+   * Example:
+   *
+   * ```cpp
+   * void in_some_method() {
+   *   publish("the/topic", 1.0);
+   *   // with two digits after the decimal point
+   *   publish("the/topic", 1.0, 2);
+   * }
+   * ```
+   *
+   * @param topic The topic to publish to.
+   * @param payload The payload to publish.
+   * @param number_decimals The number of digits after the decimal point to round to, defaults to 3 digits.
+   */
+  bool publish(const std::string &topic, float value, int8_t number_decimals = 3);
+
+  /** Publish an MQTT message with the given integer as payload.
+   *
+   * Example:
+   *
+   * ```cpp
+   * void in_some_method() {
+   *   publish("the/topic", 42);
+   * }
+   * ```
+   *
+   * @param topic The topic to publish to.
+   * @param payload The payload to publish.
+   */
+  bool publish(const std::string &topic, int value);
+
+  /** Publish a JSON-encoded MQTT message with the given Quality of Service and retain settings.
+   *
+   * Example:
+   *
+   * ```cpp
+   * void in_some_method() {
+   *   publish("the/topic", [=](JsonObject &root) {
+   *     root["the_key"] = "Hello World!";
+   *   }, 0, false);
+   * }
+   * ```
+   *
+   * @param topic The topic to publish to.
+   * @param payload The payload to publish.
+   * @param qos The Quality of Service to publish with.
+   * @param retain Whether to retain the message.
+   */
+  bool publish_json(const std::string &topic, const json::json_build_t &f, uint8_t qos, bool retain);
+
+  /** Publish a JSON-encoded MQTT message.
+   *
+   * Example:
+   *
+   * ```cpp
+   * void in_some_method() {
+   *   publish("the/topic", [=](JsonObject &root) {
+   *     root["the_key"] = "Hello World!";
+   *   });
+   * }
+   * ```
+   *
+   * @param topic The topic to publish to.
+   * @param payload The payload to publish.
+   */
+  bool publish_json(const std::string &topic, const json::json_build_t &f);
+
+  /// Check whether the MQTT client is currently connected and messages can be published.
+  bool is_connected();
+};
+
+template<typename T>
+void CustomMQTTDevice::subscribe(const std::string &topic,
+                                 void (T::*callback)(const std::string &, const std::string &), uint8_t qos) {
+  auto f = std::bind(callback, (T *) this, std::placeholders::_1, std::placeholders::_2);
+  global_mqtt_client->subscribe(topic, f, qos);
+}
+template<typename T>
+void CustomMQTTDevice::subscribe(const std::string &topic, void (T::*callback)(const std::string &), uint8_t qos) {
+  auto f = std::bind(callback, (T *) this, std::placeholders::_2);
+  global_mqtt_client->subscribe(topic, f, qos);
+}
+template<typename T> void CustomMQTTDevice::subscribe(const std::string &topic, void (T::*callback)(), uint8_t qos) {
+  auto f = std::bind(callback, (T *) this);
+  global_mqtt_client->subscribe(topic, f, qos);
+}
+template<typename T>
+void CustomMQTTDevice::subscribe_json(const std::string &topic, void (T::*callback)(const std::string &, JsonObject &),
+                                      uint8_t qos) {
+  auto f = std::bind(callback, (T *) this, std::placeholders::_1, std::placeholders::_2);
+  global_mqtt_client->subscribe_json(topic, f, qos);
+}
+template<typename T>
+void CustomMQTTDevice::subscribe_json(const std::string &topic, void (T::*callback)(JsonObject &), uint8_t qos) {
+  auto f = std::bind(callback, (T *) this, std::placeholders::_2);
+  global_mqtt_client->subscribe_json(topic, f, qos);
+}
+
+}  // namespace mqtt
+}  // namespace esphome

